src/Build/Logging/BinaryLogger/BinaryReaderWrapper.cs

```diff
@@ -0,0 +1,74 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+
+namespace Microsoft.Build.Logging
+{
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
+    using Microsoft.Build.Framework.Logging;
+    using Microsoft.Build.Shared;
+
+    /// <summary>
+    /// Implements <see cref="IBinaryReader"/> on a <see cref="System.IO.BinaryReader "/>.
+    /// </summary>
+    internal class BinaryReaderWrapper : IBinaryReader
+    {
+        private readonly BinaryReader _binaryReader;
+
+        // This is used to verify that events deserialization is not overreading expected size.
+        private readonly TransparentReadStream _readStream;
+
+        public BinaryReaderWrapper(BinaryReader binaryReader)
+        {
+            this._readStream = TransparentReadStream.EnsureTransparentReadStream(binaryReader.BaseStream);
+
+            this._binaryReader = binaryReader.BaseStream == _readStream
+                ? binaryReader
+                : new BinaryReader(_readStream);
+        }
+
+        int IBinaryReader.BytesCountAllowedToReadRemaining => _readStream.BytesCountAllowedToReadRemaining;
+
+        int? IBinaryReader.BytesCountAllowedToRead { set => _readStream.BytesCountAllowedToRead = value; }
+
+        long IBinaryReader.Position => _readStream.Position;
+
+        public const int MaxBulkRead7BitLength = 10;
+        private int[] resultInt = new int[MaxBulkRead7BitLength];
+
+        int[] IBinaryReader.BulkRead7BitEncodedInt(int numIntegers)
```

Comment:
We should probably check if the arg doesn't go over the `MaxBulkRead7BitLength` (and then either throw or just read at max `MaxBulkRead7BitLength`)

Reply:
Check added.

Reply:
Oh, I added to the buffered reader but missed this one.